pkg/datapath/linux/node.go

```diff
@@ -729,16 +743,25 @@ func (n *linuxNodeHandler) insertNeighbor(ctx context.Context, newNode *nodeType
 		nextHopIsNew = n.neighNextHopRefCount.Add(nextHopStr)
 	}
 
+	n.neighLock.Unlock() // to allow concurrent arpings bellow
+	locked = false
+
 	// nextHop hasn't been arpinged before OR we are refreshing neigh entry
+	var hwAddr net.HardwareAddr
 	if nextHopIsNew || refresh {
-		hwAddr, err := arp.PingOverLink(n.neighDiscoveryLink, srcIPv4, nextHopIPv4)
+		hwAddr, err = arp.PingOverLink(n.neighDiscoveryLink, srcIPv4, nextHopIPv4)
```

Comment:
I don't believe this is safe yet, i.e. the race detect can still complain about this. I believe you need to copy `n.neighDiscoveryLink` into a local variable while holding the `n.neighLock.Lock()`

The reason being that `n.neighDiscoveryLink` can be set more than once, i.e. when `cilium config` is called, then `NodeConfigurationChanged` is invoked again, setting `n.neighDiscoveryLink` again even if it wasn't nil before:
https://github.com/cilium/cilium/blob/e5fbe232dcbdb0ae63ba8a93a289a115e7f29afd/daemon/cmd/config.go#L101